### PR TITLE
DM-3534 DM-3561: Add Events and News Components

### PR DESCRIPTION
**Pull request by camillevilla**
### JIRA issue link
[DM-3534: Events](https://agile6.atlassian.net/browse/DM-3534?atlOrigin=eyJpIjoiZThlYjYzYWUwMTdlNDliM2FhZDk2NTk5NDRiZTUyYzQiLCJwIjoiaiJ9
)
[DM-3561: News](https://agile6.atlassian.net/browse/DM-3561?atlOrigin=eyJpIjoiYzQxOGIzOGQ3YzcxNDcxZmIwOThlMzZhY2M4MjVkYWUiLCJwIjoiaiJ9)

## Description - what does this code do?
This PR adds two new PageBuilder components with pagination: News and Events.

### Events
1. Create a new PageBuilder Group 
1. Create a new page. Set Template Type to `Narrow`.
2. Create 7 Event components
3. Save page and view page
4. Only 3 events should be displayed
5. Click 'Load more' button to load 3 more events. 
6. Click 'Load more' button to load 1 event. 
7. Event title links should have external styling and open in a new tab.

### News
1. Edit the same page
2. Create 7 News components
3. Save page and view page
4. Only 6 news items should be displayed
5. Click 'Load more' button to load 1 additional news item
6. News title links should have external styling and open in a new tab.

## Screenshots, Gifs, Videos from application (if applicable)
Screen Shot 2022-09-27 at 11 21 29 AM
Screen Shot 2022-09-27 at 11 21 17 AM

## Link to mock-ups/mock ups (image file if you have it) (if applicable)
[Events - Prototype 9.27](https://www.figma.com/file/NYiAmX7l9yGESH9sTkX7IS/XR-Drafts?node-id=1232%3A8724)
[News - Prototype 9.27](https://www.figma.com/file/NYiAmX7l9yGESH9sTkX7IS/XR-Drafts?node-id=1232%3A8771)

## Acceptance criteria
- [ ]

## Definition of done
- [x] Unit tests written (if applicable)
- [ ] e2e/accessibility tests written (if applicable)
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [x] A link has been provided to the originating JIRA issue
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

**Comment by bradjohnson92008** [comment]
Hey Camille - FYI.. your migration for page_event_components should be plural.  This one always gets me as well.. as I'm very used to table names being singular... lol.  
image

**Review by bradjohnson92008 — COMMENTED** [review]
Disregard my previous comment about the card styleing and field... i think I was looking at the wrong figma page.. but I see the latest designs in "Designs with Content" for the "News" and "Events" pages and they look good.  One thing I tested and this may be "**By Design**" is that if I add an "Event" component below a page that has "Events" first... then "News" ... I was expecting that last Event component to display under the "News" but it's added to the List of "Events" in the upper section.. again - which may be fine.. just calling that out.

**Comment by camillevilla** [comment]
@JoshingYou1 just FYI re: Brad's comment - this is by design for MVP.

> Disregard my previous comment about the card styleing and field... i think I was looking at the wrong figma page.. but I see the latest designs in "Designs with Content" for the "News" and "Events" pages and they look good. One thing I tested and this may be "**By Design**" is that if I add an "Event" component below a page that has "Events" first... then "News" ... I was expecting that last Event component to display under the "News" but it's added to the List of "Events" in the upper section.. again - which may be fine.. just calling that out.

File: app/helpers/application_helper.rb [hunk 1]
```diff
@@ -131,6 +131,12 @@ def is_internal_link?(url)
     url.start_with?('.')
   end
 
+  def get_link_classes(url)
+    classes = ["usa-link"]
+    classes << "usa-link--external" if !is_internal_link?(url)
+    classes.join(" ")
```
Comment (JoshingYou1) [inline app/helpers/application_helper.rb:137]: I think this method can be simplified quite a bit. Perhaps something like this:
```
"usa-link#{' usa-link--external' if !is_internal_link?(url)}"
```

**Review by JoshingYou1 — CHANGES_REQUESTED** [review]
Great work on this Camille! The pagination stuff is working great 😄 I left some comments, but I also have a few potential bugs to report.

1. If the user doesn't add a URL to either events or news, the element is still created and the `usa-link` and `usa-link--external` classes are added to it.


2. If the user doesn't enter a date for the `published_date`, the app breaks, producing the following error:

3. Should users be able to enter any URL they want to, without validation on at least the backend? I was able to create the link below:

File: app/controllers/page_controller.rb [hunk 2]
```diff
@@ -20,34 +24,104 @@ def show
...
+          page_param: "events-#{page_event_list_index.to_s}",
+          link_extra: "data-remote='true' class='dm-paginated-events-#{page_event_list_index}-link dm-paginated-#{page_event_list_index}-events-#{params_index.nil? ? 2 : params_index.to_i + 1}-link dm-button--outline-secondary margin-top-105 width-auto'"
+        )
+  end
+
+  def set_pagy_news_items_array(page_components)
+    page_news_item_list_index = 0 # TODO: support multiple news item lists per page
+    params_index = params["#{page_news_item_list_index}"]
+    news_items = []
+    page_components.each_with_index do |pc, index|
```
[the remark this answers is not in the document]
Comment (camillevilla) [reply app/controllers/page_controller.rb:75]: Good point! I'll get rid of those.

File: app/helpers/application_helper.rb [hunk 3]
```diff
@@ -131,6 +131,12 @@ def is_internal_link?(url)
...
+  end
```
[the remark this answers is not in the document]
Comment (camillevilla) [reply app/helpers/application_helper.rb:138]: Definitely! I'll throw one on.

**Comment by camillevilla** [comment]
@bradjohnson92008 @JoshingYou1 This PR is ready for re-review.

**Comment by camillevilla** [comment]
Got it! Thanks @JoshingYou1. Just pushed up a commit for that CSS fix.

**Review by bradjohnson92008 — APPROVED** [review]
Great Work Camille.

**Pull request merged**